core/sequencer/tracker/tracker.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracker
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/monitoring"
+	"github.com/google/trillian/util/election2"
+)
+
+const resourceLabel = "resource"
+
+var (
+	once     sync.Once
+	isMaster monitoring.Gauge
+)
+
+func createMetrics(mf monitoring.MetricFactory) {
+	isMaster = mf.NewGauge(
+		"is_master",
+		"Set to 1 for resources for which this instance is currently master",
+		resourceLabel)
+}
+
+// Tracker tracks mastership of a collection of resources.
+type Tracker struct {
+	factory     election2.Factory
+	maxHold     time.Duration
+	master      map[string]election2.Election
+	masterMu    sync.RWMutex
+	watching    map[string]bool
+	watchingMu  sync.RWMutex
+	newResource chan string
+}
+
+// New returns a new mastership tracker
+func New(factory election2.Factory, maxHold time.Duration, metricsFactory monitoring.MetricFactory) *Tracker {
+	once.Do(func() { createMetrics(metricsFactory) })
+	return &Tracker{
+		factory:     factory,
+		maxHold:     maxHold,
+		master:      make(map[string]election2.Election),
+		watching:    make(map[string]bool),
+		newResource: make(chan string),
+	}
+}
+
+// AddResource makes the mastership tracker aware of new resources.
+// The same resource may be added an unlimited number of times.
```

Comment:
Why? Is the idea that it can be added many times, but resources for monitoring it will be allocated only one time?

Reply:
yes